docs/decisions/0045-updated-vector-store-design.md

```diff
@@ -0,0 +1,811 @@
+    Task DeleteCollectionAsync(string name, CancellationToken cancellationToken = default);
+}
+
+class AzureAISearchMemoryCollectionService: IMemoryCollectionService;
+class RedisMemoryCollectionService: IMemoryCollectionService;
+class WeaviateMemoryCollectionService: IMemoryCollectionService;
+
+// Customers can inherit from our implementations and replace just the creation scenarios to match their schemas.
+class CustomerCollectionsService: AzureAISearchMemoryCollectionService, IMemoryCollectionService;
+
+// We can also create implementations that create indices based on an MLIndex specification.
+class MLIndexAzureAISearchMemoryCollectionService(MLIndex mlIndexSpec): AzureAISearchMemoryCollectionService, IMemoryCollectionService;
+
+interface IMemoryRecordService<TDataModel>
+{
+    Task<TDataModel?> GetAsync(string key, GetRecordOptions? options = default, CancellationToken cancellationToken = default);
+    Task DeleteAsync(string key, DeleteRecordOptions? options = default, CancellationToken cancellationToken = default);
+    Task<string> UpsertAsync(TDataModel record, UpsertRecordOptions? options = default, CancellationToken cancellationToken = default);
+}
+
+class AzureAISearchMemoryRecordService<TDataModel>(): IMemoryRecordService<TDataModel>;
+```
+
+#### Option 3 - Separated collection and record management with collection create separate from other operations.
+
+Vector store same as option 2 so not repeated for brevity.
+
+```cs
+
+interface IMemoryCollectionCreateService
+{
+    virtual Task CreateCollectionAsync(string name, CancellationToken cancellationToken = default);
+}
+
+class AzureAISearchChatHistoryCollectionCreateService: IMemoryCollectionCreateService;
+class AzureAISearchSemanticCacheCollectionCreateService: IMemoryCollectionCreateService;
+
+// Customers can create their own creation scenarios to match their schemas, but can continue to use our get, does exist and delete class.
+class CustomerChatHistoryCollectionCreateService: IMemoryCollectionCreateService;
+
+interface IMemoryCollectionUpdateService
+{
+    IAsyncEnumerable<string> ListCollectionNamesAsync(CancellationToken cancellationToken = default);
+    Task<bool> CollectionExistsAsync(string name, CancellationToken cancellationToken = default);
+    Task DeleteCollectionAsync(string name, CancellationToken cancellationToken = default);
+}
+
+class AzureAISearchMemoryCollectionUpdateService: IMemoryCollectionUpdateService;
+class RedisMemoryCollectionUpdateService: IMemoryCollectionUpdateService;
+class WeaviateMemoryCollectionUpdateService: IMemoryCollectionUpdateService;
+
+```
+
+#### Option 4 - Separated collection and record management with collection create separate from other operations, with collection management aggregation class on top.
+
+Variation on option 3. 
+
+```cs
+
+interface IMemoryCollectionCreateService
+{
+    virtual Task CreateCollectionAsync(string name, CancellationToken cancellationToken = default);
```

Comment:
In this option, the signature of `CreateCollectionAsync` method lacks information about collection/index schema (although, some configuration exists in Option 1: `Task CreateCollectionAsync(CollectionConfig collectionConfig, CancellationToken cancellationToken = default);`, it would be interesting to learn more about `CollectionConfig` idea.)

ADR describes how to annotate data model with attributes, whether it's a `Key` or `Vector` field etc, but we don't use this information in `CreateCollectionAsync`. Should we make this signature generic, so it will accept data model and use attributes to identify collection fields?

Below in Option 4, there is example with `abstract Task CreateCollectionAsync` and implementations: `class AzureAISearchChatHistoryMemoryService`, `class AzureAISearchSemanticCacheMemoryService` etc, so it looks like separate class per collection. 

I'm a little bit worried about this approach, because I'm not sure how it will correlate with `TDataModel` from record abstraction:
1. Every time I create a new `TDataModel` to be used in record abstraction (e.g. `Contact`, `Location`, `Message`), do I need to create a new class for collection manipulation (e.g. `AzureAISearchContactCollectionService`, `AzureAISearchLocationCollectionService` etc?
2. If at some point of time I will add a new property to `TDataModel` or rename existing one, should I replicate it in related `CollectionCreateService` class to update collection schema? If schema update is a case, should we consider adding `UpdateCollectionAsync` method as well?

It feels like collection and record abstractions should operate with custom schema in similar way. Having custom schema in record abstraction means that we can update it at any time for record operations, but these changes won't be replicated in actual collection, which may be a problem. Because I will create a collection programmatically with some schema, but as soon as I update schema in code, I will need to manually update collection schema in db. 

Let me know what you think about it. If all these scenarios are out-of-scope, we should also mention that in ADR.

Reply:
I've updated the Option 3 to clarify a bit:
```cs
// Implement a generic version of create that takes a configuation that should work for 80% of cases.
class AzureAISearchConfiguredCollectionCreateService(CollectionCreateConfig collectionConfig): IMemoryCollectionCreateService;

// Allow custom implementations of create for break glass scenarios for outside the 80% case.
class AzureAISearchChatHistoryCollectionCreateService: IMemoryCollectionCreateService;
class AzureAISearchSemanticCacheCollectionCreateService: IMemoryCollectionCreateService;
```

The thinking was to not have a create config in the method parameter as in option 1, since it'll be hard to cover all the different features from all the different dbs.  Instead, we can have a generic create implementation that takes a create config in the constructor, but there can be other create implementations too, that take different configuration.

Having a single config for the record and create has benefits and drawbacks.  On the one hand it means that you may need to add create configuration even if you don't want to use create, or if you are implementing a custom create method.  On the other hand, it avoids the issues you are raising above, where you may need to update two configs when something changes.  Let's discuss this more in person later.